src/main/java/frc/robot/subsystems/vision/VisionEstimation.java

```diff
@@ -0,0 +1,159 @@
+            VisionConstants.ROBOT_TO_PINKY);
+    private final Vision backLeftEstimator = new Vision(new PhotonCamera("Clyde"),
+            VisionConstants.ROBOT_TO_CLYDE);
+
+    private final Notifier allNotifier = new Notifier(() -> {
+        frontRightEstimator.run();
+        frontLeftEstimator.run();
+        backRightEstimator.run();
+        backLeftEstimator.run();
+    });
+
+    private OriginPosition originPosition = kBlueAllianceWallRightSide;
+
+    public VisionEstimation() {
+        this.swerveSubsystem = SwerveSubsystem.getInstance();
+
+        allNotifier.setName("runAll");
+        allNotifier.startPeriodic(0.02);
+
+        // prevTime.set(Timer.getFPGATimestamp());
+        this.swerveSubsystem.getPose();
+    }
+
+    @Override
+    public void periodic() {
+        if (VisionConstants.USE_VISION) {
+            estimatorChecker(frontRightEstimator);
+            estimatorChecker(frontLeftEstimator);
+            estimatorChecker(backRightEstimator);
+            estimatorChecker(backLeftEstimator);
+        } else {
+            allNotifier.close();
+        }
+
+        // Set the pose on the dashboard
+        var dashboardPose = swerveSubsystem.getPose();
+        if (originPosition == kRedAllianceWallRightSide) {
+            // Flip the pose when red, since the dashboard field photo cannot be rotated
+            dashboardPose = flipAlliance(dashboardPose);
+        }
+    }
+
+    /**
+    * Transforms a pose to the opposite alliance's coordinate system. (0,0) is
+    * always on the right corner of your
+    * alliance wall, the field elements are at different coordinates
+    * for each alliance.
+    * 
+    * @param poseToFlip pose to transform to the other alliance
+    * @return pose relative to the other alliance's coordinate system
+    */
+    private Pose2d flipAlliance(Pose2d poseToFlip) {
+        return poseToFlip.relativeTo(new Pose2d(
+                new Translation2d(FieldConstants.length, FieldConstants.width),
+                new Rotation2d(Math.PI)));
+    }
+
+    // public void addTrajectory(PathPlannerTrajectory traj) {
+    //   field2d.getObject("Trajectory").setTrajectory(traj);
+    // }
+
+    private Matrix<N3, N1> confidenceCalculator(EstimatedRobotPose estimation) {
+        double smallestDistance = Double.POSITIVE_INFINITY;
+        for (var target : estimation.targetsUsed) {
+            var target3D = target.getBestCameraToTarget();
+            var distance = Math
+                    .sqrt(Math.pow(target3D.getX(), 2) + Math.pow(target3D.getY(), 2) + Math.pow(target3D.getZ(), 2));
+            if (distance < smallestDistance)
+                smallestDistance = distance;
+        }
+        double poseAmbiguityFactor = estimation.targetsUsed.size() != 1
+                ? 1
+                : Math.max(
+                        1,
+                        (estimation.targetsUsed.get(0).getPoseAmbiguity()
```

Comment:
We are using `estimation.targetsUsed.get(0)` here, do we mean `target` instead? A brief comment what this code is trying to do (in addition to @Boomaa23 's suggestion to split into some variables) would help.

Reply:
No it shouldn't be target because this is changing the ambiguity factor based on the ambguity from a previouly used target, also I will try to split it into variables